### PR TITLE
Move Expression based schema project to Schema class

### PR DESCRIPTION
**Pull request by eddyxu**
Add `Schema::Project(const arrow::compute::Expression&)` method to project the fields based on the fields references in the expression.

### CODE REVIEW

File: cpp/src/lance/format/schema.cc [hunk 1]
```diff
@@ -499,6 +499,19 @@ ::arrow::Result<std::shared_ptr<Schema>> Schema::Project(
   return projection;
 }
 
+::arrow::Result<std::shared_ptr<Schema>> Schema::Project(
+    const ::arrow::compute::Expression& expr) const {
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.cc:503]: Does input need to be bound?

File: cpp/src/lance/format/schema.cc [hunk 2]
```diff
@@ -499,6 +499,19 @@ ::arrow::Result<std::shared_ptr<Schema>> Schema::Project(
...
+  if (!::arrow::compute::ExpressionHasFieldRefs(expr)) {
+    /// All scalar?
+    return nullptr;
+  }
+  std::vector<std::string> columns;
+  for (auto& ref : ::arrow::compute::FieldsInExpression(expr)) {
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.cc:509]: This does solve the nested field reference problem right?

File: cpp/src/lance/format/schema.h [hunk 3]
```diff
@@ -14,6 +14,7 @@
 
 #pragma once
 
+#include <arrow/compute/exec/expression.h>
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.h:17]: Is this Arrow public API? Or is there a higher level header file for this?

**Review by changhiskhan — APPROVED** [review]
Minor questions for clarification

File: cpp/src/lance/format/schema.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.cc:503]: It does not. This just read the schema and run projections. 
Bound expr is only needed to feed into `arrow::compute`

File: cpp/src/lance/format/schema.cc (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.cc:509]: Oh, no. That's separated.

File: cpp/src/lance/format/schema.h (hunk 3 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.h:17]: There is a `arrow/compute/api.h` should i use that one?

**Pull request merged**